README.rst
.. image:: https://img.shields.io/github/tag/thekad/pasttle?style=for-the-badge
   :target: https://github.com/thekad/pasttle/releases
   :alt: Github

.. image:: https://img.shields.io/pypi/v/pasttle?style=for-the-badge
   :target: https://pypi.python.org/pypi/pasttle
   :alt: PyPi Release

.. image:: https://img.shields.io/travis/thekad/pasttle/master?style=for-the-badge
   :target: https://travis-ci.org/thekad/pasttle
   :alt: Latest Travis CI Build

.. image:: https://img.shields.io/codecov/c/github/thekad/pasttle?style=for-the-badge
   :target: https://codecov.io/gh/thekad/pasttle
   :alt: Code Coverage

.. image:: https://img.shields.io/pypi/l/pasttle?style=for-the-badge
   :target: https://opensource.org/licenses/MIT
   :alt: License

.. image:: https://img.shields.io/pypi/pyversions/pasttle?style=for-the-badge
   :target: https://pypi.python.org/pypi/pasttle
   :alt: Python Versions

.. image:: https://img.shields.io/gitter/room/thekad/pasttle?style=for-the-badge
   :alt: Join the chat at https://gitter.im/thekad/pasttle
   :target: https://gitter.im/thekad/pasttle?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge

.. image:: https://img.shields.io/docker/build/thekad/pasttle?style=for-the-badge
   :alt: Docker build status
   :target: https://hub.docker.com/r/thekad/pasttle

Overview
========

Pasttle is based upon http://sprunge.us, born out of the necessity to:

* Have a lightweight pastebin software
* Not depend on Google AE
* Paste via CLI

Also, I wanted to play more with http://bottlepy.org

Pasttle is split in server and client, it depends on what you are trying to 
use to run one or the other.

* Server: Runs on python and needs a database (defaults to using SQLite) to
  store all the data. You want this if you are setting your own private 
  pasttle instance
* Client: Entirely written in bash. These are just thin wrappers around curl,
  you can use this to post to a pasttle server


Installing/Upgrading pasttle-server
===================================

.. note::

   If you are upgrading pasttle-server, you'd do well in reading CHANGELOG.rst
   first so you make sure the version you are upgrading has any additional
   steps.

Pasttle is part of PyPI, you can just:

.. code:: bash

    pip install pasttle

... and it should pull all the necessary components. Whether you want to install
it in a virtual environment (which I recommend) or system-wide is totally up
to you.

If you instead want to try from source code (i.e. by cloning the git repo) 
then I suggest you execute:

.. code:: bash

    python setup.py install

... and be done with it. Again, if you do this in a virtual environment or
not is up to you.

In either case, if you want to add any other database drivers on top of the
shipped SQLite driver, you need to install it separately depending on what
method you used to install pasttle to begin with.


Running the server
==================

Once you have it installed you need a database and (optionally) a different 
WSGI runner (bottle, which pasttle runs on top of, runs on wsgiref by default) 
but can run on Paste, tornado, CherryPy, and several others.

Also, since pasttle uses SQLAlchemy as the backend, you have your pick of any 
database that SQLAlchemy supports: SQLite (the default), MySQL and Postgres 
being the most famous.

Once you have your database all setup, you just need to create a pasttle.ini 
(like the one in the repo) and change the values to whatever matches your 
environment. 

After that:

.. code:: bash

    pasttle-server.py

Should start the server. If you want to use a different config file, just set 
the environment variable ``PASTTLECONF`` to the file you want to read before 
starting the server, like this:

.. code:: bash

   export PASTTLECONF=/etc/pasttle/mypasttle.ini
   pasttle-server.py
   # optionally, specify a different config section, e.g. [development]
   export PASTTLECONF=/etc/pasttle/mypasttle.ini:development
   pasttle-server.py

Alternatively, an uWSGI configuration is provided in the ``.ini`` file.
Here is a script to run the server with virtualenv option:

.. code:: bash

    #!/usr/bin/sh

    if [ -n "$VIRTUAL_ENV" ]; then
        OPT="-H $VIRTUAL_ENV"
    fi

    exec uwsgi pasttle.ini --plugin python $OPT


Available configuration options
-------------------------------

.. code:: ini

    [main]
    debug: <true/false> [default=true]
    bind: <address> [default=localhost]
    port: <port> [default=9669]
    title: <title>
    wsgi: <wsgi server to use> [default=wsgiref]
    pool_recycle: <db connection age> [default=3600]
    recent_items: <number to show on main page> [default=20]
    pygments_style: <coloration theme> [default=tango]


.. note::

    pool_recycle
            See documentation of ``sqlalchemy.create_engine`` for details
    wsgi
            WSGI server to use, look at ``bottle.server_names`` for the list

.. code:: python

    import bottle
    print(bottle.server_names.keys())
    ['cgi', 'gunicorn', 'cherrypy', 'eventlet', 'tornado', 'geventSocketIO', 'rocket', 'diesel', 'twisted', 'wsgiref', 'fapws3', 'bjoern', 'gevent', 'meinheld', 'auto', 'flup', 'gae', 'paste', 'waitress']



Running the client
==================

Running the client just requires 2 steps:

* Source pasttle.bashrc
* Run ``pasttle -h`` or ``gettle -h`` to check usage



CHANGELOG.rst
CHANGELOG
=========

v0.10.0
------

* NOTICE: Dropping support for python2 as it is no longer maintained by
  the core team (EOL on 1/1/2020)
* Bugfix: Fixed unicode support ingestion
* Enhancement: Docker image is now based on python 3.8 and ships with gunicorn
* Dev change: Using pytest as the testing framework now (with codestyle and flake8 plugins)
* Updating the dependencies to the latest stable versions from upstream
* Slight style tweaks


v0.9.5
------

* NOTICE: Dropping support for python 2.6 as it is no longer maintained by
  the core team
* Updating the dependencies to the latest stable versions from upstream


v0.9.4
------

* Bugfix: get rid of annoying cast-null warnings


v0.9.3
------

* Enhancement: adding a link to the raw version on paste view


v0.9.2
------

* Bugfix: fixing a bug on postgres for parent value
* Bugfix: fixing the copyright year


v0.9.1
------

* Bugfix: the parent value was getting set to some weird integer


v0.9
----

* DB Change: Added ``parent`` field to the ``paste`` table, integer field
* There is now a rudimentary "history" function: if you edit an entry as new,
  it will store the id of the entry it is based on, and a new link will appear
  on the final "show paste" page so you can get a diff view. This diff view is
  currently constrained to those entries without passwords
* Refined the look&feel
* Now the "show paste" page is also styled
* Reorganized the index page so the most used links appear at the top
* Paste pages now return some extra metadata that might be useful in the
  future. Comes in the form of response headers in both normal and raw mode


v0.8
----

* Improved look&feel
* Added ``lexer`` field to the table, varchar of 64


v0.7.5
------

* Split the views into separate (reusable) files


pasttle/pasttled.py
#!/usr/bin/env python
#
# -*- mode:python; sh-basic-offset:4; indent-tabs-mode:nil; coding:utf-8 -*-
# vim:set tabstop=4 softtabstop=4 expandtab shiftwidth=4 fileencoding=utf-8:
#

import bottle
from bottle.ext import sqlalchemy as sqlaplugin
try:
    import ConfigParser as configparser
except ImportError:
    import configparser
import hashlib
import logging
import os
import pygments
from pygments import formatters
from pygments import lexers
import sqlalchemy
from sqlalchemy import func
from sqlalchemy.ext import declarative
import StringIO
import sys


cfg = os.environ.get('PASTTLECONF', 'pasttle.ini').split(':')
if len(cfg) < 2:
    cfg.append('main')

cfg_file, cfg_section = cfg[0], cfg[1]

# Load configuration, or default
default_ini = StringIO.StringIO("""
[%s]
debug: true
bind: localhost
port: 9669
title: Simple paste bin
wsgi: wsgiref
pool_recycle: 3600
""" % (cfg_section,))

CONF = configparser.SafeConfigParser()
CONF.readfp(default_ini)
CONF.read(os.path.realpath(cfg_file))

DEBUG = CONF.getboolean(cfg_section, 'debug')
format = '%(asctime)s %(levelname)s %(name)s %(message)s'

# Set up logging
LOGGER = logging.getLogger('pasttle.web')
LOGGER.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
if DEBUG:
    ch.setLevel(logging.DEBUG)
else:
    ch.setLevel(logging.INFO)
formatter = logging.Formatter(format)
ch.setFormatter(formatter)
LOGGER.addHandler(ch)

# Subclass declarative base for sqla objects
Base = declarative.declarative_base()


class Paste(Base):
    """
    Main paste sqlalchemy construct for database storage
    """

    __tablename__ = 'paste'

    id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
    content = sqlalchemy.Column(sqlalchemy.Text, nullable=False)
    filename = sqlalchemy.Column(sqlalchemy.String(128))
    password = sqlalchemy.Column(sqlalchemy.String(40))
    mimetype = sqlalchemy.Column(sqlalchemy.String(64), nullable=False)
    created = sqlalchemy.Column(
        sqlalchemy.DateTime, default=func.now(), nullable=False
    )
    source = sqlalchemy.Column(sqlalchemy.String(45))

    def __init__(
        self, content, mimetype, filename=None,
        password=None, encrypt=True, source=None
    ):

        self.content = content
        self.mimetype = mimetype
        if filename and filename.strip():
            self.filename = filename.strip()[:128]
        if password:
            if encrypt:
                self.password = hashlib.sha1(password).hexdigest()
            else:
                self.password = password[:40]
        if source:
            self.source = source

    def __repr__(self):
        return u'<Paste "%s" (%s), protected=%s>' % (
            self.filename, self.mimetype, bool(self.password))

application = bottle.app()
pool_recycle = CONF.getint(cfg_section, 'pool_recycle')
LOGGER.info(
    'Recycling pool connections every %s seconds' % (pool_recycle,)
)
engine = sqlalchemy.create_engine(
    CONF.get(cfg_section, 'dsn'), echo=DEBUG,
    convert_unicode=True, logging_name='pasttle.db', echo_pool=DEBUG,
    pool_recycle=pool_recycle
)
# Create all metadata on loading, if something blows we need to know asap
Base.metadata.create_all(engine)

# Install sqlalchemy plugin
db_plugin = sqlaplugin.SQLAlchemyPlugin(engine, Base.metadata, create=True)
application.install(db_plugin)


def get_url(path=False):
    (scheme, host, q_path, qs, fragment) = bottle.request.urlparts
    if path:
        return u'%s://%s%s' % (scheme, host, q_path)
    else:
        return u'%s://%s' % (scheme, host)


@bottle.route('/')
def index():
    """
    Main index
    """

    return u"""<html>
    <head>
        <title>%(title)s</title>
        <style>
            body {
                font-family: Courier;
                font-size: 12px;
            }

            p {
                margin: 20px;
            }

            a {
                text-decoration: none;
                font-weight: bold;
            }
        </style>
    </head>
    <body>
        <pre>
pasttle(1)                          PASTTLE                          pasttle(1)

<strong>NAME</strong>
    pasttle: simple pastebin

<strong>EXAMPLES</strong>

    To post the output of a given command:

        &lt;command&gt; | curl -F "upload=&lt;-" %(url)s/post && echo

    To post the contents of a file:

        curl -F "upload=&lt;filename.ext" %(url)s/post && echo

    To post the contents of a file and force the syntax to be python:

        curl -F "upload=&lt;filename.ext" -F "syntax=python" \\
            %(url)s/post && echo

    To post the contents of a file and password protect it:

        curl -F "upload=&lt;filename.ext" -F "password=humptydumpty" \\
            %(url)s/post && echo

    You don't like sending plain-text passwords:

        curl -F "upload=&lt;filename.ext" \\
            -F "password=$( echo -n 'bootcat' | sha1sum | cut -c 1-40 )" \\
            -F "is_encrypted=yes" %(url)s/post && echo

    To get the raw contents of a paste (i.e. paste #6):

        curl %(url)s/raw/6

    To get the raw contents of a password-protected paste (i.e. paste #7):

        curl -d "password=foo" %(url)s/raw/7

    Again you don't like sending plain-text passwords:

        curl -d "is_encrypted=yes" \\
            -d "password=$( echo -n 'bootcat' | sha1sum | cut -c 1-40 )" \\
            %(url)s/raw/7

<strong>HELPERS</strong>

    There are a couple of helper functions in the link below for pasting and
    getting pastes. Import it from your ~/.bash_profile and you should be able
    to use these functions. Creating a ~/.pasttlerc helps you type less too.

    <a href="https://raw.github.com/thekad/pasttle/master/pasttle.bashrc">
        Link
    </a>

<strong>WEB FORM</strong>

    You can also use this tiny web form to easily upload content.

    <a href="%(url)s/post">
        Link
    </a>
        </pre>
        <p>Copyright &copy; Jorge Gallegos, 2012</p>
        <p>
            <a href="https://github.com/thekad/pasttle">Get the source code</a>
        </p>
    </body>
</html>
    """ % {
        'url': get_url(),
        'title': CONF.get(cfg_section, 'title'),
    }


@bottle.route('/recent')
def recent(db):
    """
    Shows an unordered list of most recent pasted items
    """

    pastes = db.query(
        Paste.id, Paste.filename, Paste.mimetype,
        Paste.created, Paste.password
    ).order_by(Paste.id.desc()).limit(20).all()
    ul = u'<ul>%s</ul>'
    li = []
    for paste in pastes:
        LOGGER.debug(paste)
        li.append(
            u'<li><a href="/%s">Paste #%d, %s (%s) %s</a></li>' %
            (
                paste.id, paste.id, paste.filename or u'',
                paste.mimetype, paste.created, )
            )
    return ul % ''.join(li)


def _edit_form(
    legend='paste new',
    content=None,
    password=None,
    syntax=None,
):
    LOGGER.debug('%s, %s, %s' % (legend, password, syntax))
    return u"""<html>
    <head>
        <title>%s</title>
        <style>
            fieldset {
                padding: 1em;
            }
            label {
                float: left;
                margin-right: 0.5em;
                padding-top: 0.2em;
                text-align: right;
                font-weight: bold;
            }
            .note {
                font-size: small;
            }
        </style>
    </head>
    <body>
        <form method="post" action="/post">
            <fieldset>
            <legend>%s</legend>
                <label for="upload">Contents: </label>
                <textarea id="upload" name="upload" rows="25"
                cols="80">%s</textarea>
                <br/>
                <label for="syntax">Force syntax: </label>
                <input id="syntax" name="syntax" value="%s" />
                <br/>
                <label for="password">Password protect this paste (40char max):
                </label>
                <input id="password" type="password" name="password"
                    maxlength="40" value="%s" />
                <br/>
                <label for="is_encrypted">Is the password encrypted? </label>
                <input type="checkbox" name="is_encrypted" %s
                id="is_encrypted" />
                <br/>
                <input type="submit" value="Submit" />
            </fieldset>
            <p class="note">
                Keep in mind that passwords are transmitted in clear-text. The
                password is not cyphered on the client-side because shipping a
                SHA1 javascript library is perhaps too much, if you check the
                "Is encrypted?" checkbox make sure your password is cyphered
                with SHA1. Perhaps you better use the readily available
            <a
            href="https://raw.github.com/thekad/pasttle/master/pasttle.bashrc">
            console helper</a>?
            </p>
        </form>
    </body>
</html>
    """ % (
        legend.capitalize(), legend.capitalize(),
        content or u'', syntax or u'', password or u'',
        u'checked="checked"' if password else u'',
    )


@bottle.route('/post')
def upload_file():
    """
    Frontend for simple posting via web interface
    """
    return _edit_form()


@bottle.post('/post')
def post(db):
    """
    Main upload interface. Users can password-protect an entry if they so
    desire. You can send an already SHA1 cyphered text to this endpoint so
    your intended password does not fly insecure through the internet
    """

    upload = bottle.request.forms.upload
    filename = None
    if bottle.request.forms.filename != '-':
        filename = bottle.request.forms.filename
    syntax = None
    if bottle.request.forms.syntax != '-':
        syntax = bottle.request.forms.syntax
    password = bottle.request.forms.password
    encrypt = not bool(bottle.request.forms.is_encrypted)
    LOGGER.debug('Filename: %s, Syntax: %s' % (filename, syntax,))
    if upload:
        if syntax:
            try:
                lexer = lexers.get_lexer_by_name(syntax)
            except lexers.ClassNotFound:
                lexer = lexers.guess_lexer(upload)
        else:
            if filename:
                try:
                    lexer = lexers.guess_lexer_for_filename(filename, upload)
                except lexers.ClassNotFound:
                    lexer = lexers.guess_lexer(upload)
            else:
                lexer = lexers.guess_lexer(upload)
        LOGGER.debug(lexer.mimetypes)
        if lexer.mimetypes:
            mime = lexer.mimetypes[0]
        else:
            mime = u'text/plain'
        source = bottle.request.remote_route
        if source:
            source = source[0]
        paste = Paste(
            content=upload, mimetype=mime, encrypt=encrypt,
            password=password, source=source, filename=filename
        )
        LOGGER.debug(paste)
        db.add(paste)
        db.commit()
        return u'%s/%s' % (get_url(), paste.id, )
    else:
        return bottle.HTTPError(400, output='No paste provided')


def _get_paste(db, id):
    """
    Queries the database for the given paste, or returns False is not found
    """

    try:
        paste = db.query(Paste).filter_by(id=id).one()
    except:
        paste = False
    return paste


def _password_protect_form():
    """
    Really simple password-protect form
    """

    return u"""<html>
    <head>
    </head>
    <body>
        <p>This entry is password protected, write in your password:</p>
        <form method="post">
            <input type="password" name="password" id="password" />
            <input type="submit" />
        </form>
    </body>
</html>
    """


def _pygmentize(paste, lang):
    """
    Guess (or force if lang is given) highlight on a given paste via pygments
    """

    if lang:
        try:
            lexer = lexers.get_lexer_by_name(lang)
        except lexers.ClassNotFound:
            lexer = lexers.get_lexer_by_name('text')
    else:
        lexer = lexers.get_lexer_for_mimetype(paste.mimetype)
    a = '<small><a href="/edit/%s">edit</a></small>' % (paste.id,)
    if paste.filename:
        title = u'%s, created on %s' % (paste.filename, paste.created, )
    else:
        title = u'created on %s' % (paste.created, )
    title = '%s %s (%s)' % (paste.mimetype, title, a,)
    LOGGER.debug(lexer)
    return pygments.highlight(
        paste.content, lexer, formatters.HtmlFormatter(
            full=True, linenos='table',
            encoding='utf-8', lineanchors='ln', title=title)
        )


@bottle.route('/<id:int>')
@bottle.post('/<id:int>')
def showpaste(db, id, lang=None):
    """
    Shows the highlighted entry on the browser. If the entry is protected
    with a password it will display a password entry and will compare against
    the database for a match
    """

    paste = _get_paste(db, id)
    if not paste:
        return bottle.HTTPError(404, output='This paste does not exist')
    password = bottle.request.forms.password
    LOGGER.debug(
        '%s == %s ? %s' % (
            hashlib.sha1(password).hexdigest(), paste.password,
            hashlib.sha1(password).hexdigest() == paste.password,
        )
    )
    if paste.password:
        if not password:
            return _password_protect_form()
        if hashlib.sha1(password).hexdigest() == paste.password:
            bottle.response.content_type = 'text/html'
            return _pygmentize(paste, lang)
        else:
            return bottle.HTTPError(401, output='Wrong password provided')
    else:
        return _pygmentize(paste, lang)


@bottle.route('/<id:int>/<lang>')
@bottle.post('/<id:int>/<lang>')
def forcehighlight(db, id, lang):
    """Forces a certain highlight against an entry"""

    return showpaste(db, id, lang)


@bottle.route('/raw/<id:int>')
@bottle.post('/raw/<id:int>')
def showraw(db, id):
    """
    Returns the plain-text version of the entry. If the entry is protected
    with a password it will display a simple password entry form until the
    password is a match in the database
    """

    paste = _get_paste(db, id)
    if not paste:
        return bottle.HTTPError(404, output='This paste does not exist')
    password = bottle.request.forms.password
    is_encrypted = bool(bottle.request.forms.is_encrypted)
    if not is_encrypted:
        match = hashlib.sha1(password).hexdigest()
    else:
        match = password
    LOGGER.debug(
        '%s == %s ? %s' % (match, paste.password, match == paste.password, )
    )
    if paste.password:
        if not password:
            return _password_protect_form()
        if match == paste.password:
            bottle.response.content_type = 'text/plain'
            return paste.content
        else:
            return bottle.HTTPError(401, output='Wrong password provided')
    else:
        bottle.response.content_type = 'text/plain'
        return paste.content


@bottle.post('/edit/<id:int>')
@bottle.route('/edit/<id:int>')
def edit(db, id):
    """
    Edits the entry. If the entry is protected with a password it will display
    a simple password entry form until the password is a match in the database
    """

    paste = _get_paste(db, id)
    if not paste:
        return bottle.HTTPError(404, output='This paste does not exist')
    password = bottle.request.forms.password
    is_encrypted = bool(bottle.request.forms.is_encrypted)
    if not is_encrypted:
        match = hashlib.sha1(password).hexdigest()
    else:
        match = password
    LOGGER.debug(
        '%s == %s ? %s' % (
            match, paste.password,
            match == paste.password,
        )
    )
    kwargs = {
        'password': paste.password,
        'content': paste.content,
        'syntax': lexers.get_lexer_for_mimetype(paste.mimetype).aliases[0],
    }
    if paste.password:
        if not password:
            return _password_protect_form()
        if match == paste.password:
            return _edit_form('edit entry #%s' % (paste.id,), **kwargs)
        else:
            return bottle.HTTPError(401, output='Wrong password provided')
    else:
        return _edit_form('edit entry #%s' % (paste.id, ), **kwargs)


def main():
    LOGGER.info('Using Python %s' % (sys.version, ))
    bottle.run(
        application, host=CONF.get(cfg_section, 'bind'),
        port=CONF.getint(cfg_section, 'port'), reloader=DEBUG,
        server=CONF.get(cfg_section, 'wsgi')
    )

if __name__ == '__main__':
    sys.exit(main())
